src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,164 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int count;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (tail != null) {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        if (head == null) {
+            head = newNode;
+        }
+        count++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexToAdd(index);
+        if (index == size()) {
+            add(value);
+        } else if (index == 0) {
+            addToHead(value);
+        } else {
+            Node<T> source = getNodeByIndex(index);
+            Node<T> toAdd = new Node<>(source.previous, value, source);
+            source.previous.next = toAdd;
+            source.previous = toAdd;
+            count++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        list.forEach(this::add);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexToGet(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexToGet(index);
+        Node<T> toSet = getNodeByIndex(index);
+        T oldValue = toSet.value;
+        toSet.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexToGet(index);
+        Node<T> toRemove = getNodeByIndex(index);
+        unlink(toRemove);
+        return toRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        return unlink(getNodeByValue(object));
     }
 
     @Override
     public int size() {
-        return 0;
+        return count;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
+    }
+
+    private void checkIndexToGet(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private void checkIndexToAdd(int index) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException("Invalid index");
```

Comment:
Maybe in this case will be better to add a more informative message about exception. And also you can add the corresponding index to the message in both methods.

Reply:
It makes sense. Fixed =)